File: client/src/services/MessageService.js
import api from "../utils/axiosConfig";

const MessageService = {

  // =========================================================
  // SEND MESSAGE USING REST
  // WebSocket will normally be used for real-time messaging
  // =========================================================

  async sendMessage(receiverEmail, content, taskId = null) {
    const response = await api.post("/messages", {
      receiverEmail,
      content,
      taskId,
    });

    return response.data;
  },

  // =========================================================
  // GET CONVERSATION WITH A USER
  // =========================================================

  async getConversation(email) {
    const response = await api.get(
      "/messages/conversation",
      {
        params: {
          email,
        },
      }
    );

    return response.data;
  },

  // =========================================================
  // GET ALL CONVERSATIONS
  // =========================================================

  async getConversations() {
    const response = await api.get(
      "/messages/conversations"
    );

    return response.data;
  },

  // =========================================================
  // MARK CONVERSATION AS READ
  // =========================================================

  async markAsRead(senderEmail) {
    await api.put("/messages/read", null, {
      params: {
        senderEmail,
      },
    });
  },

  // =========================================================
  // GET TOTAL UNREAD COUNT
  // =========================================================

  async getUnreadCount() {
    const response = await api.get(
      "/messages/unread-count"
    );

    return response.data.count;
  },
};

export default MessageService;

File: server/src/main/java/com/campusmarketplace/server/service/MessageService.java
package com.campusmarketplace.server.service;

import com.campusmarketplace.server.dto.request.SendMessageRequest;
import com.campusmarketplace.server.dto.response.MessageResponse;

import java.util.List;

public interface MessageService {

    // REST message sending
    MessageResponse sendMessage(
            SendMessageRequest request
    );

    // WebSocket message sending
    MessageResponse sendMessage(
            SendMessageRequest request,
            String senderEmail
    );

    List<MessageResponse> getConversation(
            String otherUserEmail
    );

    List<String> getConversations();

    void markConversationAsRead(
            String senderEmail
    );

    long getUnreadCount();
}